### PR TITLE
6️⃣ Lecture 6: Introduction to zero-knowledge

### PR DESCRIPTION
**Pull request by ZamDimon**
# Objective

This pull request adds the 6th lecture on zero knowledge. Namely, we consider:
- Proofs in cryptography.
- Defining zero-knowledge and argument of knowledge.
- Fiat-Shamir Transformation

### CODE REVIEW

File: lectures/6-intro-zk.tex [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+\begin{itemize}
+    \item \textbf{Zero-knowledge:} The prover wants to convince the verifier that the statement is true without revealing the witness. 
+    \item \textbf{Argument of knowledge:} Moreover, typically we want to make sure that the verifier, besides the statement correctness, ensures that the prover \textbf{knows} such a witness related to the statement.
+    \item \textbf{Succinctness:} The proof should be short, ideally logarithmic in the size of the statement. This is crucial for practical applications, especially in the blockchain space where we cannot allow to publish long proofs on-chain.
```
Comment (Alex-distributed-lab) [inline lectures/6-intro-zk.tex]: Verification efficiency also matters

File: lectures/6-intro-zk.tex [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+\end{itemize}
+
+\begin{example}
+    Suppose, given a hash function\footnote{The notation $\{0,1\}^*$ means binary strings of arbitrary length} $H: \{0,1\}^* \to \{0,1\}^{\ell}$, the prover $\mathcal{P}$ wants to convince the verifier $\mathcal{V}$ that he knows the preimage $x \in \{0,1\}^*$ such that $H(x) = y$ for some given public value $y \in \{0,1\}^{\ell}$. The properties listed above are interpreted as follows:
+    \begin{itemize}
+        \item \textbf{Zero-knowledge:} The prover $\mathcal{P}$ does not want to reveal \textit{anything} about the pre-image $x$ to the verifier $\mathcal{V}$.
+        \item \textbf{Argument of knowledge:} Given a string $y \in \{0,1\}^{\ell}$ it is not sufficient for a prover to merely state that $y$ has a pre-image. The prover $\mathcal{P}$ must demonstrate to a verifier $\mathcal{V}$ that he \textbf{knows} such a pre-image $x \in \{0,1\}^*$.
+        \item \textbf{Succinctness:} If the hash function takes $n$ operations to compute, the proof should be \textbf{much} shorter than $n$ operations. State-of-art solutions can provide proofs that are $O(\log n)$ is size!
```
Comment (Alex-distributed-lab) [inline lectures/6-intro-zk.tex]: Proof size and how many operations you must perform for verification are slightly different)

File: lectures/6-intro-zk.tex [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    \end{itemize}
+\end{example}
+
+\subsection{Relations and Languages}
+Recall that \textbf{relation} $\mathcal{R}$ is just a subset of $\mathcal{X} \times \mathcal{Y}$ for two arbitrary sets $\mathcal{X}$ and $\mathcal{Y}$. Now, we are going to intrduce the notion of a \emph{language of true statements} based on $\mathcal{R}$.
```
Comment (Alex-distributed-lab) [inline lectures/6-intro-zk.tex]: introduce

File: lectures/6-intro-zk.tex [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+\end{remark}
+
+\subsubsection{The Verifier's View}
+Suppose that the interaction between $\mathcal{V}$ and $\mathcal{P}$ has ended with the successful verification. What has $\mathcal{V}$ learned? Well, first things first, he has learned that the statement is true, that is $x \in \mathcal{L}_{\mathcal{R}}$. However, he has also learned something more: he has learned the transcript of the interaction, that is the sequence of messages between $\mathcal{P}$ and $\mathcal{V}$.
+
+\begin{definition}
+    Interaction between $\mathcal{P}$ and $\mathcal{V}$ consists of verifier's queries $(q_1,q_2,\dots,q_{\ell})$, prover's responses $(m_1,m_2,\dots,m_{\ell})$, and $\mathcal{V}$'s random coins $(r_1,r_2,\dots,r_{\ell})$. The view of the verifier $\mathcal{V}$, denoted as $\mathsf{view}_{\mathcal{V}}(\mathcal{P},\mathcal{V})[x]$, is a random variable $(q_1,m_1,r_1,q_2,m_2,r_2,\dots,q_{\ell},m_{\ell},r_{\ell})$ that is determined by the random coins of $\mathcal{V}$ and the messages of $\mathcal{P}$ after the interaction with the statement $x$. See Figure below.
+\end{definition}
+
+\begin{figure}[H]
```
Comment (Alex-distributed-lab) [inline lectures/6-intro-zk.tex:210]: It is a bit confusing following the definition

**Review by Alex-distributed-lab — APPROVED** [review]

File: lectures/6-intro-zk.tex (hunk 2 above)
Comment (ZamDimon) [reply lectures/6-intro-zk.tex]: Yeah, formally, "number of operations" typically means "number of gates in the arithmetical circuit", but I believe "number of operations" is informal enough for the purposes of introduction :)

**Pull request merged**